### PR TITLE
[docs] Update sphinx to 3.5.4

### PR DESCRIPTION
**Pull request by garymm**
Sphinx 4.x is out, but it seems that requires many more changes to
adopt. So instead use the latest version of 3.x, which includes
several nice features.

* Add some noindex directives to deal with warnings that would otherwise
  be triggered by this change due to conflicts between the docstrings
  declaring a function and the autodoc extension declaring the
  same function.
* Update distributions.utils.lazy_property to make it look like a
  regular property when sphinx autodoc inspects classes.

### CODE REVIEW

File: torch/distributions/utils.py [hunk 1]
```diff
@@ -92,7 +92,7 @@ def probs_to_logits(probs, is_binary=False):
     return torch.log(ps_clamped)
 
 
-class lazy_property(object):
+class lazy_property:
```
Comment (albanD) [inline torch/distributions/utils.py:95]: This transforms it into an old style class. We never want to use old style classes no?

Comment (garymm) [reply torch/distributions/utils.py:95]: There is no such thing as an old-style class in Python 3, and PyTorch doesn't support Python 2, so writing `(object)` has no effect.

Comment (albanD) [reply torch/distributions/utils.py:95]: Ho I missed that point when we moved out of Python 2!

**Comment by garymm** [comment]
@albanD I can haz approval + import plz?

**Review by albanD — APPROVED** [review]
LGTM

**Comment by garymm** [comment]
@albanD can you please merge this?

**Comment by albanD** [comment]
Hey!
I have been trying but this is breaking quite a lot of internal tests consistently (not sure why yet).
Could you actually do a rebase on top of master here to try and see if that helps?

**Comment by garymm** [comment]
Rebased

**Comment by albanD** [comment]
Ho we see the same failures here now :D So that was a valid race of something that landed in the meantime?

**Comment by garymm** [comment]
Reproduced the failures locally, fixed them, and pushed a new version.

**Comment by albanD** [comment]
Awesome!
Thanks for the fix.
I'll try to land that now!

**Comment by albanD** [comment]
@garymm while looking closer at the final doc, there are some un-expected spacing in the function definitions.
For example, 
image
Becomes
image

With a lot of extra spacing and in particular between the nn and the Conv2d.

Do you think we can fix that?

**Comment by garymm** [comment]
I see what you mean. I guess there's something that can be changed in the [theme](https://github.com/pytorch/pytorch_sphinx_theme) CSS, but I really have no experience with CSS.
@holly1238 @Bgreen23 any suggestions?

In the mean time I'll take a look to see if I can figure it out on my own.

If we need to revert, let me know and I will revert just the requirements.txt change and leave the other parts of this PR.

**Comment by NicolasHug** [comment]
For ref, I opened an issue about this a while ago: https://github.com/pytorch/pytorch_sphinx_theme/issues/115
https://github.com/pytorch/pytorch_sphinx_theme/issues/118 is also tangentially related

**Comment by albanD** [comment]
@garymm I would love to be able to keep this new version in. I think it is ok for the master doc to not look optimal for a few days while we investigate this.

**Comment by garymm** [comment]
Trying to fix this in https://github.com/pytorch/pytorch_sphinx_theme/pull/134

**Pull request closed without merging**